internal/setup/setup.go

```diff
@@ -0,0 +1,212 @@
+// Call [Reset] to reset the task's state and re-prompt the user.
+func ConfirmRun(ctx context.Context, key string, task Task, prompt string) error {
+	if prompt == "" {
+		return taskError(key, redact.Errorf("empty confirmation prompt"))
+	}
+	return run(ctx, key, task, prompt)
+}
+
+var defaultPrompt = func(msg string) (response any, err error) {
+	err = survey.AskOne(&survey.Confirm{Message: msg}, &response)
+	return response, err
+}
+
+func run(ctx context.Context, key string, task Task, prompt string) error {
+	state := loadState(key)
+	if !task.NeedsRun(ctx, state.LastRun) {
+		return nil
+	}
+
+	oldState, newState := state, &state
+	defer func() {
+		if oldState != *newState {
+			saveState(key, *newState)
+		}
+	}()
+
+	if prompt != "" {
+		state.ConfirmPrompt.Message = prompt
+		if state.ConfirmPrompt.Asked && !state.ConfirmPrompt.Allowed {
+			// We've asked before and the user said no.
+			return taskError(key, ErrUserRefused)
+		}
+
+		resp, err := defaultPrompt(prompt)
+		if err != nil {
+			return taskError(key, redact.Errorf("prompt for confirmation: %v", err))
+		}
+		state.ConfirmPrompt.Asked = true
+		state.ConfirmPrompt.Allowed, _ = resp.(bool)
+		if !state.ConfirmPrompt.Allowed {
+			return taskError(key, ErrUserRefused)
+		}
+	}
+
+	state.LastRun = RunInfo{
+		Time:    time.Now(),
+		Version: build.Version,
+	}
+	if err := task.Run(ctx); err != nil {
+		state.LastRun.Error = err.Error()
+		return taskError(key, err)
+	}
+	return nil
+}
+
+// Reset removes a task's state so that it acts as if it has never run.
+func Reset(key string) {
+	err := os.Remove(statePath(key))
+	if errors.Is(err, os.ErrNotExist) {
+		return
+	}
+	if err != nil {
+		err = taskError(key, fmt.Errorf("remove state file: %v", err))
+		debug.Log(err.Error())
+	}
+}
+
+type state struct {
+	ConfirmPrompt confirmPrompt `json:"confirm_prompt,omitempty"`
+	LastRun       RunInfo       `json:"last_run,omitempty"`
+}
+
+type confirmPrompt struct {
+	Message string `json:"message"`
+	Asked   bool   `json:"asked"`
+	Allowed bool   `json:"allowed"`
+}
+
+func loadState(key string) state {
+	path := statePath(key)
+	b, err := os.ReadFile(path)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			err = taskError(key, fmt.Errorf("load state file: %v", err))
+			debug.Log(err.Error())
+		}
+		return state{}
+	}
+	loaded := state{}
+	if err := json.Unmarshal(b, &loaded); err != nil {
+		err = taskError(key, fmt.Errorf("load state file %s: %v", path, err))
+		debug.Log(err.Error())
+		return state{}
+	}
+	return loaded
+}
+
+func saveState(key string, s state) {
+	path := statePath(key)
+	data, err := json.MarshalIndent(s, "", "  ")
+	if err != nil {
+		err = taskError(key, fmt.Errorf("save state file: %v", err))
+		debug.Log(err.Error())
+		return
+	}
+
+	err = os.WriteFile(path, data, 0o644)
+	if errors.Is(err, os.ErrNotExist) {
+		err = os.MkdirAll(filepath.Dir(path), 0o755)
+		if err == nil {
+			err = os.WriteFile(path, data, 0o644)
+		}
+	}
```

Comment:
`MkdirAll` doesn't error if path already exists or if only partial path exists so you can simplify:

```golang 
err := os.MkdirAll(filepath.Dir(path), 0o755)
if err == nil {
  err = os.WriteFile(path, data, 0o644)
}
```